platform/common/cog-gl-utils.c

```diff
@@ -0,0 +1,83 @@
+/*
+ * cog-gl-utils.c
+ * Copyright (C) 2021 Igalia S.L.
+ *
+ * Distributed under terms of the MIT license.
+ */
+
+#include "cog-gl-utils.h"
+
+#include "../../core/cog.h"
+
+void
+cog_gl_shader_id_destroy(CogGLShaderId *shader_id)
+{
+    if (shader_id && *shader_id) {
+        glDeleteShader(*shader_id);
+        *shader_id = 0;
+    }
+}
+
+CogGLShaderId
+cog_gl_shader_id_steal(CogGLShaderId *shader_id)
+{
+    g_assert(shader_id);
+    CogGLShaderId result = *shader_id;
+    *shader_id = 0;
+    return result;
+}
+
+CogGLShaderId
+cog_gl_load_shader(const char *source, GLenum kind, GError **error)
+{
+    g_assert(source != NULL);
+    g_assert(kind == GL_VERTEX_SHADER || kind == GL_FRAGMENT_SHADER);
+
+    g_auto(CogGLShaderId) shader = glCreateShader(kind);
+    glShaderSource(shader, 1, &source, NULL);
+
+    GLenum err;
+    if ((err = glGetError()) != GL_NO_ERROR) {
+        g_set_error_literal(error, COG_PLATFORM_EGL_ERROR, err, "Cannot set shader source");
+        return 0;
+    }
+
+    glCompileShader(shader);
+    if ((err = glGetError()) != GL_NO_ERROR) {
+        g_set_error_literal(error, COG_PLATFORM_EGL_ERROR, err, "Cannot compile shader");
+        return 0;
+    }
+
+    GLint ok = GL_FALSE;
+    glGetShaderiv(shader, GL_COMPILE_STATUS, &ok);
+    if (ok == GL_TRUE)
+        return cog_gl_shader_id_steal(&shader);
+
+    char buffer[4096] = {
+        '\0',
+    };
+    glGetShaderInfoLog(shader, G_N_ELEMENTS(buffer), NULL, buffer);
+    g_set_error(error, COG_PLATFORM_EGL_ERROR, 0, "Shader compilation: %s", buffer);
+    return 0;
+}
+
+bool
+cog_gl_link_program(GLuint program, GError **error)
+{
+    glLinkProgram(program);
+
+    GLint status = 0;
+    glGetProgramiv(program, GL_LINK_STATUS, &status);
+    if (status)
+        return true;
+
+    GLint log_length = 0;
+    glGetProgramiv(program, GL_INFO_LOG_LENGTH, &log_length);
+    {
+        char log[log_length + 1];
```

Comment:
Yes, variable length arrays (VLAs) have [made it to C99](https://en.wikipedia.org/wiki/Variable-length_array#C99). They are quite handy IMO, but I am open to change this if people think it's a bad idea to use them.

Reply:
VLAs are considered somewhat harmful for a number of reasons, given that they're basically the equivalent of alloca() and could use an unbounded amount of stack memory without any way to detect that. In this case, it's an error path so I don't think it would hurt anything to do a heap allocation.

Reply:
Sure, I can turn this into a `g_new()` and allocate the array on the heap. Will do!

Reply:
Actually, I am also going to replace the static 4096-byte buffer used to get the shader compilation log in `cog_gl_load_shader()` to retrieve the log length with `glGetShaderiv()` and allocate a buffer dynamiically.